rfcs/2022-04-09-accessibility.md

```diff
@@ -0,0 +1,505 @@
+
+- [Focus order](#keyboard-tab-order)
+- [Focus ring styling](#focus-ring-styling)
+- [Focus traps](#focus-trap)
+  - Focus restoration
+- Page transitions
+
+## Focus order
+
+Focus order or tab order we already covered in [keyboard tab order section](#keyboard-tab-order)
+
+## Focus ring styling
+
+The focus ring must be visible to all users, 
+By default, the browser uses the user agent specific focus styling. But this behaviour can be overriden with CSS. 
+
+Best practices for focus styling:
+
+- Provide focus styles that are highly visible.
+- Make sure that a visible element has focus at all times when using a keyboard.
+- Avoid using `*:focus { outline: none }` snippet to hide focus
+- Design focus rings such that it has proper contrast
+
+> Not providing visible focus ring is [Failure of Success Criterion 2.4.7](https://www.w3.org/WAI/WCAG21/Techniques/failures/F78)
+
+**Keyboard only focus rings**
+
+While focus indicators are neccesary,  
+Mouse clicks also changes the focus, for example clicking on a button also makes the focus ring visible.  
+
+This behavior can be undesirable from design perspective, preferring focus styling to only be present only if a keyboard is used can be a solution. 
+
+<img width="45%" src="./images/accessibility/focus-visible-demo.gif" alt="Demo of focus visible css pesudo class, showing two buttons and comparing the results" />
+
+**Implementations**
+
+- CSS
+- ReactAria
+
+**CSS:**  
+
+> [Demo](https://codesandbox.io/s/blade-a11y-css-focus-ring-051xff?file=/src/App.tsx)
+
+CSS provides us with a pesudo class called [`:focus-visible`](https://css-tricks.com/keyboard-only-focus-styles). Browser support for focus-visible is [good enough.](https://caniuse.com/?search=%3Afocus-visible).  
+And we can also use the [official pollyfill](https://github.com/WICG/focus-visible) for older browsers.
+
+**ReactAria:**  
+
+> [Demo](https://codesandbox.io/s/blade-a11y-reactaria-focus-ring-hc3cub?file=/src/App.tsx)
+
+ReactAria provides a [FocusRing](https://react-spectrum.adobe.com/react-aria/FocusRing.html) component which solves this. 
+
+While this works great, this solution can be a bit overkill since react aria does this all with javascript and have it's own event handling system.  
+I think it will be better and easier if we `use the platform™` 
+
+## Focus trap
+
+> [Specification]()
+
+Trapping focus is a behaviour we usually want when there is modality in a page. 
+
+**Why focus traps?**
+
+Focus traps are essential to communicate a modal's bounds to groups of people who are visually impaired or have mobility issues, And helps inform what is and is not modal content.
+
+The idea is that if for part of the site we prevent clicks, we should also prevent focus.
+
+- People with low vision who rely on screen readers need to know when the modal opens.
+- People with low vision who cannot properly see the focus ring can get lost if focus isn't contained in the modal, This can be confusing and disorienting.
```

Comment:
```suggestion
- People with low vision who cannot properly see the focus ring can get lost if focus isn't contained in the modal, this can be confusing and disorienting.
```